pyccel/codegen/printing/cwrappercode.py

```diff
@@ -110,21 +141,25 @@ def _print_FunctionDef(self, expr):
         keyword_list_name = self.get_new_name(used_names,'kwlist')
         keyword_list = PyArgKeywords(keyword_list_name, arg_names)
 
+        code = ''
         wrapper_body = [keyword_list]
         wrapper_body_translations = []
 
         parse_args = []
         # TODO: Simplify (to 1 line?)
         # TODO: Handle optional args
         for a in expr.arguments:
-            collect_type, cast_func = self.get_PyArgParseType(a.dtype)
+            collect_var, cast_func = self.get_PyArgParseType(used_names, a)
             if cast_func is not None:
                 # TODO: Add other properties
-                collect_var = Variable(dtype=collect_type,
-                        name=self.get_new_name(used_names, a.name+"_tmp"))
+                cast_func = self.pop_cast_function(cast_func)        
+                if not cast_func in self._cast_functions_set:
+                    code += self._print(cast_func)
+                    self._cast_functions_set.add(cast_func)
                 wrapper_vars.append(collect_var)
                 parse_args.append(collect_var)
-                wrapper_body_translations.append(cast_func(a, collect_var))
+                cast_func_call = FunctionCall(cast_func, [collect_var])
+                wrapper_body_translations.append(AliasAssign(a, cast_func_call))
```

Comment:
Why do you use an AliasAssign?

Reply:
`AliasAssign` is necessary when `a` is a pointer. This is the case when `a` is a `PyccelObject`. However it should not be used systematically otherwise we risk having incorrect code (e.g.)
```
def f(a = None):
    print(a)
```
will have a signature:
```
void f(_Bool* a);
```
Calling `AliasAssign(a, cast_func_call)` will give:
```
_Bool* a;
a = pyint_to_bool(a_tmp);
```
instead of :
```
_Bool* a;
*a = pyint_to_bool(a_tmp);
```

To know whether we need to use Assign or AliasAssign we need to know the return type of cast_func. The simplest place to check this is in `get_cast_function`, `get_PyArgParseType` or `get_PyBuildValue`. These functions would therefore return the expression required to cast to the correct type as before rather than returning the function that will be used to translate the expression

Reply:
ok I understand